src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -470,14 +471,26 @@ private async Task RefreshIndividualKeyValues()
 
         private async Task RefreshKeyValueAdapters()
         {
-            foreach (IKeyValueAdapter adapter in _options.Adapters)
+            if (!AdapterRefreshSemaphore.Wait(0))

```

Comment:
If we don't need a lock we shouldn't take it.

It looks like we may want
```
if (adapters.Any(a => a.NeedsRefresh() &&
    AdapterRefreshSemaphore.Wait(0))
{
    try
    {
        if (adapters.Any(a => a.NeedsRefresh())
        {
            await SetData(_applicationSettings).ConfigureAwait(false);
         }
    }
    finally
    {
           AdapterRefreshSemaphore.Release();
    ]
}
```

Reply:
Done